### PR TITLE
[Perf] Batch GPU->CPU sync for multi-item scoring embeddings

### PR DESCRIPTION
**Pull request by fortunecookiee**
## Motivation

In the multi-item scoring classification / reward path the pooler returns one 2D tensor per request, so `BatchResultProcessor._convert_embeddings` takes the `else` branch and calls `.tolist()` once per tensor:

```python
embeddings = [tensor.tolist() for tensor in embeddings]
```

Each `.tolist()` is a separate device-to-host copy that synchronizes the CUDA stream, so a batch of N requests pays N syncs at the end of every prefill. The syncs serialize against the next forward pass, and the cost scales linearly with batch size — it shows up as a growing gap between forward end and response send at high concurrency.

## Modifications

Add `batch_convert_2d_tensors_to_lists` to `srt/managers/utils.py`: concatenate the per-request tensors along dim 0, do **one** `.tolist()`, then re-split on the host using the recorded row counts. This turns N syncs into 1.

The helper falls back to the original per-tensor `.tolist()` when the input is empty or shape-irregular (differing last dim — e.g. per-request matryoshka truncation), so it is never wrong, just not batched, in those cases.

Only the non-sparse, list-of-tensors branch changes. The sparse branch and the single pre-stacked `torch.Tensor` branch are untouched.

## Accuracy Tests

Output is bit-identical to the previous code path by construction (same tensors, same order, only the transfer is batched). Verified equivalence against the per-tensor `[t.tolist() for t in tensors]` reference on: uniform batches, single-tensor batches, zero-row tensors, ragged column counts (fallback path), empty input, and fp16 tensors — all exact matches.

## Benchmarking and Profiling

The win is N-1 fewer D2H syncs per prefill batch in the multi-item scoring embedding path; it grows with batch size and is a no-op at batch size 1.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/developer_guide/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/developer_guide/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/developer_guide/accuracy_evaluation.html).
- [x] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34619397169](https://github.com/sgl-project/sglang/actions/runs/34619397169)
Latest PR Test (Extra): :x: [Run #34619396722](https://github.com/sgl-project/sglang/actions/runs/34619396722)
Latest PR Test (AMD ROCm 10): :x: [Run #34619396863](https://github.com/sgl-project/sglang/actions/runs/34619396863)

### CODE REVIEW

**Comment by fortunecookiee** [comment]
Rebased onto current `main` and **reworked the approach** after a self-review found the original version optimized the wrong layer.

The previous version batched `.tolist()` inside `SchedulerBatchResultProcessor._convert_embeddings`. But with overlap scheduling — the default — `EmbeddingBatchResult.copy_to_cpu()` has already run and `copy_done.synchronize()` has already fired before `_convert_embeddings` is reached, so the tensors are already host-resident. There were no D2H syncs left to batch there, and the added `torch.cat` was pure extra work on the default path.

The batching now lives in `copy_to_cpu`, where the N pinned allocations and N async copies actually exist, and applies to both the `embeddings` and `pooled_hidden_states` lists. `_can_batch_d2h` gates it on CUDA sources plus uniform dtype/device/rank/trailing-shape so the concat is exactly invertible by `torch.split`; everything else (notably per-request matryoshka truncation) falls back to the per-tensor path.
